tests/Authenticator.hs

```diff
@@ -0,0 +1,426 @@
+
+      -- 2. Let credentialOptions be a new empty set of public key credential
+      -- sources.
+      -- 3. If allowCredentialDescriptorList was supplied, then for each
+      -- descriptor of allowCredentialDescriptorList:
+      -- 3.1. Let credSource be the result of looking up descriptor.id in this authenticator.
+      -- 3.2. If credSource is not null, append it to credentialOptions.
+      -- 4. Otherwise (allowCredentialDescriptorList was not supplied), for each
+      -- key -> credSource of this authenticator’s credentials map, append
+      -- credSource to credentialOptions.
+      -- 5. Remove any items from credentialOptions whose rpId is not equal to rpId.
+      let credentialOptions = filter
+            (\(_, o) -> pkcsRpId o == rpId)
+            $ case allowCredentialDescriptorList of
+              Just descriptors -> mapMaybe (authenticatorLookupCredential authenticator . M.pkcdId) descriptors
+              Nothing -> Map.assocs aCredentials
+
+      -- 6. If credentialOptions is now empty, return an error code equivalent
+      -- to "NotAllowedError" and terminate the operation.
+      when (null credentialOptions) (fail "NotAllowedError: No CredentialOptions (None of the existing credentials were deemed acceptable)")
+
+      -- 7. Prompt the user to select a public key credential source
+      -- slectedCredential from credentialOptions. Collect an authorization
+      -- gesture confirming user consent for using selectedCredential. The
+      -- prompt for the authorization gesture may be shown by the authenticator
+      -- if it has its own output capability, or by the user agent otherwise.
+
+      -- If requireUserVerification is true, the authorization gesture MUST
+      -- include user verification.
+
+      -- If requireUserPresence is true, the authorization gesture MUST include
+      -- a test of user presence.
+
+      -- If the user does not consent, return an error code equivalent to
+      -- "NotAllowedError" and terminate the operation.
+      -- NOTE: We always assume the user cooperates for now and choses the
+      -- first possible source
+      let (selectedKey, selectedCredential) = head credentialOptions
+
+      -- 8. Let processedExtensions be the result of authenticator extension
+      -- processing for each supported extension identifier → authenticator
+      -- extension input in extensions.
+      -- TODO: We don't suport extensions
+
+      -- 9. Increment the credential associated signature counter or the global
+      -- signature counter value, depending on which approach is implemented by
+      -- the authenticator, by some positive value. If the authenticator does
+      -- not implement a signature counter, let the signature counter value
+      -- remain constant at zero.
+      let updatedCounter = pkcsCounter selectedCredential + 1
+      let credentials = Map.update (\v -> Just $ v {pkcsCounter = updatedCounter}) selectedKey aCredentials
```

Comment:
You can use `Map.adjust` instead

Reply:
Nice, I didn't know that existed. HLS should give a hint for that.